### PR TITLE
Fix layer_normalize gradients

### LINKED ISSUE #2902: [Bug]: Wrong gradients in the CUDA implementation of Layer Norm
### What Operating System(s) are you seeing this problem on?

Linux (x86-64)

### dlib version

19.24

### Python version

N/A

### Compiler

GCC 12.3.0

### Expected Behavior

Expect the tests to pass for LayerNorm, both in CPU and CUDA implementations.

### Current Behavior

This test passed (I have CUDA enabled of course):
https://github.com/davisking/dlib/blob/46e59a2174228922d19d2887756d6dbfef80dc04/dlib/test/dnn.cpp#L603-L653

In the first part (before the `#if DLIB_USE_CUDA`) we are checking that the Layer Normalization actually does what it says on CPU: normalizes each sample.

In the second part, I compute the CUDA version and check it is equal to the CPU version.
That works, Then, I proceed to compute the gradients on both CPU and GPU and check if they are equal.

Both these tests pass. However, this one only passes on CPU and does not pass on GPU:
https://github.com/davisking/dlib/blob/46e59a2174228922d19d2887756d6dbfef80dc04/dlib/test/dnn.cpp#L2007-L2012

I have stared at both implementations for a while and I can't see what I am doing wrong.
If someone could have an extra look, I'll appreciate that.

### Steps to Reproduce

Just run the test suite with CUDA enabled.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by arrufat**
Closes #2902

### CODE REVIEW

**Comment by arrufat** [comment]
It still doesn't fix the discrepancy between GPU and CPU, but fixes a bug in the implementation.

**Comment by davisking** [comment]
I'm looking at this and not sure what the code is supposed to be doing.    Go through the contracts and make sure they are right.  Like 
```
    void layer_normalize (
        const double eps, 
        resizable_tensor& dest,
        resizable_tensor& means,
        resizable_tensor& invstds,
        const tensor& src, 
        const tensor& gamma,
        const tensor& beta 
    );   
    /*!  
        requires
            - eps > 0
            - src.num_samples() == gamma.size() == beta.size()
            - have_same_dimensions(gamma, beta) == true
            - beta.num_samples() ==beta.nr() ==gamma.nc() == 1
```

That's saying beta and gamma are the same shape and all the dimensions are 1 except `k`, which would have to be `src.num_sample()`.  So by that what was in the code before this would make sense.  But running some of these I see that `gamma` and beta don't have the shape that requires clause says they do.  So there is inconsistency in how these are being interpreted.  I.e. is gamma.size() == src.num_samples() or is it `src.k() * src.nr() * src.nc()`?

That's totally at the root of the problem here.  Everything starts with having contracts that are right.  Then put DLIB_ASSERT statements that check all the requires statements so you know for sure they are not being violated.  That will chase down the problem.  Although you've got to decide what the arguments are first.  Not sure what you want them to be for this layer.  I would think `gamma.size() == src.num_samples()` as that's probably the most typical variant of layer norm though.

**Comment by davisking** [comment]
Inside `layer_normalize_gradient()` for the cuda code, there are also local `resiable_tensor` variables (`dvars` and `dmeans`), those can't live there.  It's real expensive to be creating and destroying tensors.  They all need to live in a layer object so they aren't created and destroyed when the network runs.  But rather are allocated once.  But more than that kernels run asynchronously in cuda, so that `_cuda_layer_normalize_gradient` kernel launches but then those two variables are immediately freed.  That's probably why the cuda version isn't working.  Since it's running on dangling pointers.

Sorry I never really looked at this.  You write such good PRs I just kinda skimmed this one and was like "yeah another Adria PR, going to be great and looks great :+1: :D "

**Comment by arrufat** [comment]
Oh, right, what was I thinking. It looks like I got confused half-way through the code, where I should normalize each channel independently, but I ended up trying to normalize along `k`, `nr`, and `nc`. Definitely, `k` should not be included in the normalization.

Hopefully, I will fix that and the dangling pointer tonight if life allows.

**Comment by arrufat** [comment]
I was just checking this again:
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html

image

It seems like, it does normalize along C, H, W. and there's one gamma and one beta for each normalized item.

```python
import torch
N, C, H, W = 20, 5, 10, 10
x = torch.randn(N, C, H, W)
layer_norm = torch.nn.LayerNorm([C, H, W])
y = layer_norm(x)
sum(p.numel() for p in layer_norm.parameters() if p.requires_grad)  # 1000 = 2 * (5 * 10 * 10) 
```

So, maybe the issue is just the dangling pointers? I will make sure the contracts are correct and respected, though.

**Comment by davisking** [comment]
Yeah, might be just the dangling pointer and everything else is fine.

**Comment by arrufat** [comment]
I am now confident about the CPU implementation, however, the CUDA version still fails.
There's a mismatch between CPU and CUDA and in the beta and gamma gradients, which are trivial, but I can't spot the mistake.
It's getting late, I'll try again later. Feel free to check, it must be something really stupid.

**Comment by arrufat** [comment]
I honestly don't know what else to do.
The CPU version works correctly now, but no the CUDA version.

However, if you run the `test_layer_normalize` with CUDA enabled, you'll see that all the functionality is on par with the CPU version:
- normalized output tensor
- src_grad
- gamma_grad
- beta_grad
- dmeans
- dvars

All of them are within the tolerance error of 1e-5 or 1e-4. However, `test_layer` with `layer_norm` still fails.

**Comment by arrufat** [comment]
Ok, fixed a race condition, now `test_layer` complains like this:
```
Average parameter gradient error is somewhat large at: 0.00713434                                                                                                                                                                                                                                                               
```

**Comment by davisking** [comment]
Nice. I'm away from my computer. I'll look in a bit. Seems like you got it 🥳

**Comment by arrufat** [comment]
It took an awful amount of time...

File: dlib/cuda/cuda_dlib.cu [hunk 1]
```diff
@@ -2201,8 +2168,73 @@ namespace dlib
...
+            {
+                const auto ps = s + n * ks * num;
+                const auto pgi = gi + n * ks * num;
+                const float invstd_pow = -0.5 * std::pow(v[n], 3.0f);
+                float temp_dv = 0;
+                for (auto i : grid_stride_range(0, ks * num))
+                {
+                    const float x_hat = (ps[i] - m[n]) * v[n];
+                    atomicAdd(&bg[i / num], pgi[i]);
+                    atomicAdd(&gg[i / num], pgi[i] * x_hat);
```
Comment (arrufat) [inline dlib/cuda/cuda_dlib.cu:2203]: Not too happy about this. I will fix it tonight.

It makes the backward pass awfully slow.
In my experiment, a network trains at about 320 img/s, if I comment out these two lines, it trains at about 2560 img/s (similar rate as the batch norm)

File: dlib/cuda/cuda_dlib.cu [hunk 2]
```diff
@@ -2201,8 +2168,88 @@ namespace dlib
...
+                for (auto i : grid_stride_range(0, num))
+                {
+                    const float x_hat = (ps[i] - m[n]) * v[n];
+                    temp_bg += pgi[i];
+                    temp_gg += pgi[i] * x_hat;
+                }
+                warp_reduce_atomic_add(bg[k], temp_bg);
+                warp_reduce_atomic_add(gg[k], temp_gg);
+            }
+            __syncthreads();
```
[the remark this answers is not in the document]
Comment (davisking) [reply dlib/cuda/cuda_dlib.cu:2225]: Yeah that kind of warp reduction loop is the best way I know to do it too.

**Comment by arrufat** [comment]
After the previous two commits, the network went to train from 320 img/s to 2450 img/s (close to the official CUDA/CUDNN batch norm at 2560 img/s)

**Comment by davisking** [comment]
> After the previous two commits, the network went to train from 320 img/s to 2450 img/s (close to the official CUDA/CUDNN batch norm at 2560 img/s)

Yeah that's awesome.   All the tests are passing for me too on a GPU machine.  Passing for you too now?  Anything else you want to change before I merge it? :D

**Comment by arrufat** [comment]
No, I think it's done now. FINALLY.

**Comment by davisking** [comment]
Yeah nice, thanks for all the good work.  Looks perfect :D
